### ISSUE TITLE
cancancan does not correctly use table aliases in the WHERE clause

### ISSUE THREAD

**Issue by kxmbrian** [Coursemology/coursemology2#706]
When we define the following rule
```
read_hash = { course_user: { course: { course_users: { user_id: user.id, workflow_state: 'approved', role: [1,2,3] } } } }
can :read, Course::ExperiencePointsRecord, read_hash
```
Calling `accessible_by` generates the query:

```sql
...
FROM
	"course_experience_points_records" 
		LEFT OUTER JOIN "course_users"
			ON "course_users" . "id" = "course_experience_points_records" . "course_user_id" 
		LEFT OUTER JOIN "courses"
			ON "courses" . "id" = "course_users" . "course_id"
				AND "courses" . "instance_id" = $1 
		LEFT OUTER JOIN "course_users" "course_users_courses"
			ON "course_users_courses" . "course_id" = "courses" . "id"
WHERE
	"course_experience_points_records" . "course_user_id" = $2
	AND (
		"course_users" . "user_id" = $3
		AND "course_users" . "workflow_state" = $4
		AND "course_users" . "role" IN (1, 2, 3)
	)  [["instance_id", 1], ["course_user_id", 4], ["user_id", 9], ["workflow_state", "approved"]]
```
which is incorrect.  `course_users` should be generated as `course_users_courses` in the WHERE clause:
```sql
WHERE
	"course_experience_points_records" . "course_user_id" = $2
	AND (
		"course_users_courses" . "user_id" = $3
		AND "course_users_courses" . "workflow_state" = $4
		AND "course_users_courses" . "role" IN (1, 2, 3)
	)  [["instance_id", 1], ["course_user_id", 4], ["user_id", 9], ["workflow_state", "approved"]]
```

**Comment by kxmbrian** [comment]
## tl;dr

Any issues if I invest time to modify (overhaul?) cancancan to combine conditions by merging scopes?

## Stuff tried

The above issue arises because cancancan only extracts the leaf values of the hash `read_hash` to construct the `WHERE` clause.

Since this query works:

```ruby
Course::ExperiencePointsRecord.
includes(course_user: { course: :course_users}).
where(course_user: { course: { course_users: { user_id: 1, workflow_state: 'approved', role: [1,2,3] } } } )
```

an obvious approach would be to get cancancan to do something similar: pass the whole hash into `where` instead of passing in the conjunction (`AND`) of the leaf nodes. 

The main challenge is that there might be more than one rule per model. Cancancan takes the conjunction of *all* the conditions. These are either hashes or escaped SQL (cancancan disallows combining conditions if a scope is used as a condition). The current approach is to convert each condition into an SQL string, glue them together with `AND`, and then only plonk them into `where`. 

I don't yet have any good solution that works with this manual `AND`-ing. If we follow this approach, we have to make sure that the right SQL string gets generated for each condition. In this case, the SQL string generator needs to take into account repeated tables. Some not so ideal ways:

- Put the hash into `where` and generate the query for each condition. Extract the `WHERE` clauses and combine them. This is problematic on a few counts. 

	- Manipulating raw SQL is error-prone.

	- We need to include the `joins` or `includes` for the aliasing to be correct.
	
	- Reflection is not done on the models, as a result, the association names are taken to be table names. In general, queries generated this way will not run successfully.

Illustration:

```
Course::ExperiencePointsRecord.
where(course_user: { randomz: { course_users: { user_id: 1, workflow_state: 'approved', role: [1,2,3] } } } ).
to_sql
```

gives

```
SELECT "course_experience_points_records".* 
FROM "course_experience_points_records" 
WHERE ("course_users"."user_id" = 1 AND 
    "course_users"."workflow_state" = 'approved' AND 
    "course_users"."role" IN (1, 2, 3))
```

- Another way is improve cancancan functions to detect and handle repeated tables. The function extracting the tables to be joined also needs to be upgraded to handle the aliasing. That would be a huge mess compared to what it is doing now, which is just to pass a hash to `includes`. Since AR/EL seems to be doing this already, I think this might be re-inventing the wheel.

In the ideal case, we should not have to manipulate SQL strings wherever avoidable. Instead of manually gluing them with `AND`, perhaps a more feasible way might be to create and merge scopes. If need be, we could use AREL directly. 

This change will be comparatively major. Not sure if there are some other straight-forward way I'm missing. Any comments on the scope merging / AREL way?

**Comment by lowjoel** [comment]
Why do we need to worry about repeated tables? If we have two different hashes from two different rules, we just .where(rule_1_hash).where(rule_2_hash) and the resulting query is the conjunction of both?

**Comment by kxmbrian** [comment]
Sorry, cancancan takes the conjunction of the conditions in a rule, but takes the disjunction of the conditions (which are conjunctions if the rule has more than one condition) from each rule.

Look likes merging scopes will not help because they AND also, besides the fact we can just chain `where`s.

**Comment by lowjoel** [comment]
AR does expose an .or on all relations in Rails 5. Otherwise, there's a way to convert the AR::Relation to an Arel condition, and those can be ORed.

http://stackoverflow.com/questions/7976358/activerecord-arel-or-condition is a bit old, might need some extra googling.

**Comment by kxmbrian** [comment]
Are we going to migrate to Rails 5? If yes, when?

**Comment by lowjoel** [comment]
Not within the next 3 months. Maybe second half this year. Quite a few gems need to be updated.
